logic/Rule.java

```diff
@@ -472,42 +472,118 @@ public Relation(RelationConstraint relation, IsaConstraint isa, Rule rule) {
+                        playerEntry.putIfAbsent(player, 1);
                     });
+                });
+                return relationMap.equals(insertedMap);
+            }
+
+            private FunctionalIterator<com.vaticle.typedb.core.concept.thing.Relation> matchRelation(
+                    RelationType relationType, ConceptMap whenConcepts,
+                    TraversalEngine traversalEng, ConceptManager conceptMgr) {
+                Identifier.Variable.Retrievable relationId = relation().owner().id();
+                RelationTraversal traversal = new RelationTraversal(relationId, set(relationType.getLabel())); // TODO include inheritance
+                relation().players().forEach(rp -> {
+                    Identifier.Variable.Retrievable playerId = rp.player().id();
+                    assert rp.roleType().isPresent() && rp.roleType().get().label().isPresent()
+                            && whenConcepts.contains(playerId);
+                    traversal.player(playerId, whenConcepts.get(playerId).asThing().getIID(),
+                                     set(roleType(rp, relationType, whenConcepts).getLabel())); // TODO include inheritance
+                });
+                return traversalEng.relations(traversal).map(conceptMgr::conceptMap)
+                        .map(conceptMap -> conceptMap.get(relationId).asRelation());
+            }
+
+            private RelationType relationType(ConceptMap whenConcepts, ConceptManager conceptMgr) {
+                if (isa().type().reference().isName()) {
+                    Reference.Name typeReference = isa().type().reference().asName();
+                    assert whenConcepts.contains(typeReference) && whenConcepts.get(typeReference).isRelationType();
+                    return whenConcepts.get(typeReference).asRelationType();
                 } else {
-                    return insert(relationType, whenConcepts, conceptMgr);
+                    assert isa().type().reference().isLabel();
+                    return conceptMgr.getRelationType(isa().type().label().get().label());
+                }
+            }
+
+            private RoleType roleType(RelationConstraint.RolePlayer rp, RelationType scope, ConceptMap whenConcepts) {
+                if (rp.roleType().get().reference().isName()) {
+                    return whenConcepts.get(rp.roleType().get().reference().asName()).asRoleType();
+                } else {
+                    assert rp.roleType().get().reference().isLabel();
+                    return scope.getRelates(rp.roleType().get().label().get().properLabel().name());
                 }
             }
 
-            private FunctionalIterator<Map<Identifier.Variable, Concept>> insert(RelationType relationType, ConceptMap whenConcepts, ConceptManager conceptMgr) {
+            private Map<Identifier.Variable, Concept> insert(RelationType relationType, ConceptMap whenConcepts) {
```

Comment:
`insert` and `thenConcept` are very similar, but calling `thenConcepts` after `insert` would incur real cost as the role types would be retrieved from the graph twice.